ai/integrations/autogen/tests/test_toolkit.py

```diff
@@ -0,0 +1,274 @@
+    mock_executor._wrap_function.assert_called_once_with(sample_autogen_tool.fn)
+    mock_executor.register_function.assert_called_once_with(
+        {sample_autogen_tool.name: "wrapped_function"}
+    )
+
+
+@requires_databricks
+@pytest.mark.parametrize("use_serverless", [True, False])
+def test_toolkit_e2e(use_serverless, monkeypatch):
+    monkeypatch.setenv(USE_SERVERLESS, str(use_serverless))
+    client = get_client()
+    with set_default_client(client), create_function_and_cleanup(client, schema=SCHEMA) as func_obj:
+        toolkit = UCFunctionToolkit(function_names=[func_obj.full_function_name])
+        tools = toolkit.tools
+        assert len(tools) == 1
+        tool = tools[0]
+        assert func_obj.full_function_name.replace(".", "__") in tool.description
+        assert func_obj.comment in tool.description
+        assert tool.client_config == client.to_dict()
+
+        input_args = {"code": "print(1)"}
+        result = json.loads(tool.fn(**input_args))["value"]
+        assert result == "1\n"
+
+        toolkit = UCFunctionToolkit(function_names=[f"{CATALOG}.{SCHEMA}.*"])
+        assert len(toolkit.tools) >= 1
+        assert func_obj.tool_name in [t.name for t in toolkit.tools]
+
+
+@requires_databricks
+@pytest.mark.parametrize("use_serverless", [True, False])
+def test_toolkit_e2e_manually_passing_client(use_serverless, monkeypatch):
+    monkeypatch.setenv(USE_SERVERLESS, str(use_serverless))
+    client = get_client()
+    with set_default_client(client), create_function_and_cleanup(client, schema=SCHEMA) as func_obj:
+        toolkit = UCFunctionToolkit(function_names=[func_obj.full_function_name], client=client)
+        tools = toolkit.tools
+        assert len(tools) == 1
+        tool = tools[0]
+        assert tool.name == func_obj.tool_name
+        assert func_obj.full_function_name.replace(".", "__") in tool.description
+        assert func_obj.comment in tool.description
+        assert (
+            "code: 'Python code to execute. Remember to print the final result to stdout.'"
+            in tool.description
+        )
+        assert tool.client_config == client.to_dict()
+        input_args = {"code": "print(1)"}
+        result = json.loads(tool.fn(**input_args))["value"]
+        assert result == "1\n"
+
+        toolkit = UCFunctionToolkit(function_names=[f"{CATALOG}.{SCHEMA}.*"], client=client)
+        assert len(toolkit.tools) >= 1
+        assert func_obj.tool_name in [t.name for t in toolkit.tools]
+
+
+@requires_databricks
+@pytest.mark.parametrize("use_serverless", [True, False])
+def test_multiple_toolkits(use_serverless, monkeypatch):
+    monkeypatch.setenv(USE_SERVERLESS, str(use_serverless))
+    client = get_client()
+    with set_default_client(client), create_function_and_cleanup(client, schema=SCHEMA) as func_obj:
+        toolkit1 = UCFunctionToolkit(function_names=[func_obj.full_function_name])
+        toolkit2 = UCFunctionToolkit(function_names=[f"{CATALOG}.{SCHEMA}.*"])
```

Comment:
Is a wildcard valid here? I don't believe it is. 
We don't have logic to handle this (we shouldn't implement this and we have explicit logic in the client to raise a ValueError if a wildcard is submitted) and as such, the method call `client.get_function(<fn name>)` raises:

https://github.com/unitycatalog/unitycatalog/blob/d0c1aa2621f6fb6eecde6bf651679d8cde76f9d7/ai/core/src/unitycatalog/ai/core/databricks.py#L443-L447

You could make this functionality work by using a `client.list_functions()` call and then passing in the list of all functions that are registered, but for the purposes of the test, we should probably just create two separate functions (we have another test handler for creating a function with a different registration name) or just use the same func_obj for two separate toolkit instantiations. 

For reference, here is your stack trace on the integration test:

```sh
_________________________ test_multiple_toolkits[True] _________________________

use_serverless = True
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7f94e8f8a020>

    @requires_databricks
    @pytest.mark.parametrize("use_serverless", [True, False])
    def test_multiple_toolkits(use_serverless, monkeypatch):
        monkeypatch.setenv(USE_SERVERLESS, str(use_serverless))
        client = get_client()
        with set_default_client(client), create_function_and_cleanup(client, schema=SCHEMA) as func_obj:
            toolkit1 = UCFunctionToolkit(function_names=[func_obj.full_function_name])
>           toolkit2 = UCFunctionToolkit(function_names=[f"{CATALOG}.{SCHEMA}.*"])

integrations/autogen/tests/test_toolkit.py:144: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/unitycatalog/ai/autogen/toolkit.py:120: in validate_toolkit
    self.tools_dict = process_function_names(
core/src/unitycatalog/ai/core/utils/function_processing_utils.py:175: in process_function_names
    tools_dict[f.full_name] = uc_function_to_tool_func(
/opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/unitycatalog/ai/autogen/toolkit.py:161: in uc_function_to_autogen_tool
    tool = pydantic_function_tool(
/opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/openai/lib/_tools.py:43: in pydantic_function_tool
    "parameters": to_strict_json_schema(model),
/opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/openai/lib/_pydantic.py:18: in to_strict_json_schema
    schema = model_json_schema(model)
/opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/openai/_compat.py:172: in model_json_schema
    return model.model_json_schema()
/opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/pydantic/main.py:476: in model_json_schema
    return model_json_schema(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'pydantic.main.BaseModel'>, by_alias = True
ref_template = '#/$defs/{model}'
schema_generator = <class 'pydantic.json_schema.GenerateJsonSchema'>
mode = 'validation'

    def model_json_schema(
        cls: type[BaseModel] | type[PydanticDataclass],
        by_alias: bool = True,
        ref_template: str = DEFAULT_REF_TEMPLATE,
        schema_generator: type[GenerateJsonSchema] = GenerateJsonSchema,
        mode: JsonSchemaMode = 'validation',
    ) -> dict[str, Any]:
        """Utility function to generate a JSON Schema for a model.
    
        Args:
            cls: The model class to generate a JSON Schema for.
            by_alias: If `True` (the default), fields will be serialized according to their alias.
                If `False`, fields will be serialized according to their attribute name.
            ref_template: The template to use for generating JSON Schema references.
            schema_generator: The class to use for generating the JSON Schema.
            mode: The mode to use for generating the JSON Schema. It can be one of the following:
    
                - 'validation': Generate a JSON Schema for validating data.
                - 'serialization': Generate a JSON Schema for serializing data.
    
        Returns:
            The generated JSON Schema.
        """
        from .main import BaseModel
    
        schema_generator_instance = schema_generator(by_alias=by_alias, ref_template=ref_template)
    
        if isinstance(cls.__pydantic_core_schema__, _mock_val_ser.MockCoreSchema):
            cls.__pydantic_core_schema__.rebuild()
    
        if cls is BaseModel:
>           raise AttributeError('model_json_schema() must be called on a subclass of BaseModel, not BaseModel itself.')
E           AttributeError: model_json_schema() must be called on a subclass of BaseModel, not BaseModel itself.

/opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/pydantic/json_schema.py:2277: AttributeError
```

What's going on here is:
1. We pass an invalid name in, a ValueError is raised. 
2. The function object definition will be defined as a BaseModel in Pydantic during the Pydantic model creation stage since it's an invalid definition. 
3. The validation will fail as it attempts to parse the json representation of the model since this isn't allowed in Pydantic (can't operate on a BaseModel instance with that API) 

I'm going to file a PR to update this validation logic in the other flavors, but in the meantime, let's not use wildcards in the test definitions :)

Reply:
Understood. I think the previous version allowed wildcard but I have updated the test to use `client.list_functions()`